internal/overlord/logstate/syslog.go

```diff
@@ -0,0 +1,207 @@
+	pid            string
+	structuredData string // TODO: use this for future log labels
+
+	// connection info
+	conn          net.Conn
+	address       *url.URL
+	waitReconnect time.Duration
+	closed        bool
+
+	data bytes.Buffer
+}
+
+func newSyslogClient(addr string) (*syslogClient, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "-"
+	}
+
+	u, err := url.Parse(addr)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + addr)
+	}
+	if err != nil {
+		return nil, err
+	} else if u.Scheme != "tcp" && u.Scheme != "udp" && u.Scheme != "" {
+		return nil, fmt.Errorf("invalid syslog server address scheme %q", u.Scheme)
+	}
+	if u.Scheme == "" {
+		u.Scheme = "tcp"
+	}
+
+	return &syslogClient{
+		version:        1, // RFC5424
+		pid:            "-",
+		host:           hostname,
+		msgid:          "-",
+		priority:       priorityVal(FacilityUserLevelMessage, SeverityInformational),
+		structuredData: syslogStructuredData("pebble", canonicalPrivEnterpriseNum, nil),
+		address:        u,
+	}, nil
+}
+
+// Syslog Priority values - see RFC 5424 6.2.1
+const (
+	// Facility values
+	FacilityUserLevelMessage = 1
+
+	// Severity values
+	SeverityInformational = 6
+)
+
+// priorityVal calculates the syslog Priority value (PRIVAL) from the given
+// Facility and Severity values. See RFC 5424, sec 6.2.1 for details.
+func priorityVal(facility, severity int) int {
+	return facility*8 + severity
+}
+
+// syslogStructuredData formats the given labels into a structured data section
+// for a syslog message, according to RFC5424 section 6.
+func syslogStructuredData(name string, enterpriseNum int, labels map[string]string) string {
+	if len(labels) == 0 {
+		return "-"
+	}
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "[%s@%d", name, enterpriseNum)
+
+	// Sort label keys to get deterministic output
+	keys := make([]string, 0, len(labels))
+	for key := range labels {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	for _, key := range keys {
+		fmt.Fprintf(&buf, " %s=\"", key)
+		value := labels[key]
+		// escape the value according to RFC5424 6.3.3
+		for i := 0; i < len(value); i++ {
+			// don't use "for _, c := range value" as we don't want runes
+			c := value[i]
+			if c == '"' || c == '\\' || c == ']' {
+				buf.WriteByte('\\')
+			}
+			buf.WriteByte(c)
+		}
+		buf.WriteByte('"')
+	}
+	buf.WriteByte(']')
+	return buf.String()
+}
+
+func (c *syslogClient) encodeEntry(entry servicelog.Entry) {
+	// TODO: optimise memory allocations here
+	// format defined by RFC 5424
+	timestamp := entry.Time.Format(time.RFC3339)
+	msg := fmt.Sprintf("<%d>%d %s %s %s %s %s %s %s",
+		c.priority, c.version, timestamp, c.host, entry.Service,
+		c.pid, c.msgid, c.structuredData, entry.Message)
+
+	// Octet framing as per RFC 5425.
+	framed := fmt.Sprintf("%d %s", len(msg), msg)
+	c.data.Write([]byte(framed))
+}
+
+// Send messages to remote syslog server.
+func (c *syslogClient) Send(entries []servicelog.Entry) error {
```

Comment:
If we're doing the `if len(entries) == 0 { return nil }` check in Loki, we should do it here as well for consistency.